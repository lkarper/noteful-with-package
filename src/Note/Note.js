import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import NotesContext from '../NotesContext/NotesContext';
import config from '../config';

class Note extends Component {

    state = {
        note: {},
        error: null,
    }

    static contextType = NotesContext;

    static defaultProps = {
        match: {
            params: {
                noteId: '',
            },
        },
        history: {
            push: () => {},
        }
    }

    deleteNoteRequest = (noteId, cb) => {
        fetch(`${config.API_URL}/notes/${noteId}`, {
            method: 'DELETE',
            headers: {
                'content-type': 'application/json',
                'Authorization': `Bearer ${config.API_KEY}`
            },
        })
        .then(response => {
            if (!response.ok) {
                return response.json().then(error => {
                    throw error;
                });
            }
        })
        .then(data => {
            cb(noteId);
            this.props.history.push('/');
        })
        .catch(error => {
            console.log(error);
            this.setState({ error: error.message });
        });
    }

    componentDidUpdate(prevProps, prevState) {
        if (!Object.keys(prevState.note).length && this.context.notes.length && !prevState.error) {
            const note = this.context.notes.find(n => n.id === parseInt(this.props.match.params.noteId));
            if (note) {
                this.setState({
                    note,
                    error: null,
                });
            } else {
                this.setState({
                    error: 'Could not find note by id.  Check the URL for typos and try again.'
                })
            }
        } 
    }

    componentDidMount() {
        if (this.context.notes.length) {
            const note = this.context.notes.find(n => n.id === parseInt(this.props.match.params.noteId));
            if (note) {
                this.setState({
                    note,
                    error: null,
                });
            } else {
                this.setState({
                    error: 'Could not find note by id.  Check the URL for typos and try again.'
                });
            }
        }
    }

    render() {
        const error = this.state.error;
        const errorHTML = (
            <div>
                <h2>Could not delete note: {error}.</h2>
                <p>Check your connection and try again.</p>
            </div>
        );

        const { note } = this.state;

        if (Object.keys(note).length) {
            return (
                <section className="notes">
                    <h2>{note.name}</h2>
                    <p>{`Last modified on: ${Date(note.date_modified).toString()}`}</p>
                    <button 
                        type="button"
                        onClick={() => this.deleteNoteRequest(note.id, this.context.deleteNote)}    
                    >Delete Note</button>
                    <Link
                        to={{
                            pathname: `/note/${this.props.match.params.noteId}/update-note`,
                        }}
                    >Edit Note</Link>
                    {error ? errorHTML : ''}
                    <p>{note.content}</p>
                </section>
            );
        }

        return (
            <section className="notes">
                {this.state.error || this.context.noteError ?
                    <> 
                        <h2>Error: {this.state.error || this.context.noteError}</h2>
                        <p>Check your connection and reload the page.</p>
                    </>
                    : <h2>Loading...</h2>
                }
            </section>
        );
    }
}

Note.propTypes = {
    history: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired,
}

export default Note;