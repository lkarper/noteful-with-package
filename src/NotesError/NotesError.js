import React, { Component } from 'react';

class NotesError extends Component {
    state ={
        hasError: false,
    };

    static getDerivedStateFromError(error) {
        return { hasError: true };
    }

    render() {
        if (this.state.hasError) {
            return <h2>Cound not display notes.  Check your connection and try again.</h2>;
        }
        return this.props.children;
    }
}

export default NotesError;