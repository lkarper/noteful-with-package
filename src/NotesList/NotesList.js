import React from 'react';
import { Link } from 'react-router-dom';
import ShortNote from '../ShortNote/ShortNote';
import NotesContext from '../NotesContext/NotesContext';

const NotesList = () => {
    return (
        <NotesContext.Consumer>
            {value => {
                const error = value.noteError;
                const errorHTML = (
                    <li className="notes-load-error">
                        <h2>Sorry, could not fetch notes from server: {error}.</h2>
                        <p>Check your connection and reload the page.</p>
                    </li>
                );

                const notes = value.notes.map(note => 
                    <ShortNote key={note.id} note={note} />
                );

                return (
                    <section className="notes">
                        {!error && !value.notes.length ? <h2>Loading...</h2> :
                            <>
                                <ul>
                                    {notes.length && !error ? notes : <li>No notes saved.</li>}
                                    {error ? errorHTML : ''}
                                </ul>
                                <Link 
                                    className="add-note-button"
                                    to={{
                                        pathname: "/add-note",
                                        state: {
                                            folderId: null,
                                        },
                                    }}
                                >
                                    Add note
                                </Link>
                            </>
                        }
                    </section>
                );
            }}
        </NotesContext.Consumer>
    );
}

export default NotesList;