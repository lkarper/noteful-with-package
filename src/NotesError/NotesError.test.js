import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import NotesError from './NotesError';
import NotesList from '../NotesList/NotesList';
import Note from '../Note/Note';

describe.skip('NotesError component', () => {

    it('renders the UI without crashing', () => {
        const wrapper = mount(
            <NotesError>
                <BrowserRouter>
                    <NotesList />
                </BrowserRouter>
            </NotesError>
        );
        wrapper.unmount();
    });

    it('renders the UI as expected', () => {
        const wrapper = mount(
            <NotesError>
                <BrowserRouter>
                    <NotesList />
                </BrowserRouter>
            </NotesError>
        );
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    it('renders an error message when an error occurs in the DOM tree beneath it', () => {
        const wrapper = mount(
            <NotesError>
                <BrowserRouter>
                    <Note location={""} />
                </BrowserRouter>
            </NotesError>
        );
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

});