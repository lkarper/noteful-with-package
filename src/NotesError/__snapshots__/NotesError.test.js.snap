// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesError component renders an error message when an error occurs in the DOM tree beneath it 1`] = `
<NotesError>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Note
        history={
          Object {
            "push": [Function],
          }
        }
        location=""
        match={
          Object {
            "params": Object {
              "noteId": "",
            },
          }
        }
      >
        <section
          className="notes"
        >
          <h2>
            Loading...
          </h2>
        </section>
      </Note>
    </Router>
  </BrowserRouter>
</NotesError>
`;

exports[`NotesError component renders the UI as expected 1`] = `
<NotesError>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <NotesList>
        <section
          className="notes"
        >
          <h2>
            Loading...
          </h2>
        </section>
      </NotesList>
    </Router>
  </BrowserRouter>
</NotesError>
`;
